### PR TITLE
build: Unified Docker Image

### PR DESCRIPTION
**Pull request by tnightengale**
**PR Checklist**

- [x] A description of the changes is added to the description of this PR.
- [x] If there is a related issue, make sure it is linked to this PR.
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added or modified a feature, documentation in `docs` is updated

**Description of changes**

This PR simplifies the use of Docker by merging the cli and server into a unified, multi-arch compatible, Dockerfile with clean syntax, and brings the project closer to releasing an official image on DockerHub. See #319, #482, and #480.

**Additional info**
- Example of the proposed GitHub Actions workflow [successfully building and pushing images](https://github.com/tnightengale/unitycatalog/actions/runs/10930448242/job/30343362272) on my fork
- Shaved off 1GB from the `./Dockerfile` image to get it down to about 1.35GB

This PR does the following:
- Simplifies the main `Dockerfile` and enables arm64 native builds
- Add `compose.yaml` to allow running UC with one command `docker compose up -d`
- Adds standard build and push GitHub Actions workflow. Actually pushing the official image to DockerHub can be enabled in a subsequent PR.

### CODE REVIEW

**Comment by tnightengale** [comment]
@dennyglee I have cycles this week to try and push towards releasing an official image. Would love your feedback here to collaborate on this.

**Comment by dennyglee** [comment]
Thanks @tnightengale !  I do like the streamlined flow but let me also include @jaceklaskowski as he's been deeply involved with the docker creation.

File: ui/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+COPY package.json yarn.lock ./
+
+# TODO: configure this directly from the ui app
+# Work around to set the proxy host; required for running locally with docker-compose
+RUN <<EOF
+set -ex
+apt-get update
+apt-get install -y jq
+jq --arg proxy "http://$PROXY_HOST:8080" '.proxy = $proxy' package.json > temp.json
```
Comment (tnightengale) [inline ui/Dockerfile:15]: Hey 👋 @jamieknight-db! I am trying to parameterize the UI react app to use the api url `http://backend:8080` instead of `http:localhost:8080` when running inside of docker compose. My react knowledge is pretty dismal; is there a better way to configure the react app than just replacing the `proxy` key inside `package.json`?

Comment (tnightengale) [reply ui/Dockerfile:15]: For context, this PR enables us to spin up the frontend and backend with just `docker compose up -d`

File: build.sbt [hunk 2]
```diff
@@ -270,6 +270,14 @@ lazy val server = (project in file("server"))
       "com.h2database" %  "h2" % "2.2.224",
 
       "org.hibernate.orm" % "hibernate-core" % "6.5.0.Final",
+      // For loading properties from the environment
+      "uk.org.webcompere" % "lightweight-config" % "1.3.0",
+
+      // For MySQL backend
+      "com.mysql" % "mysql-connector-j" % "8.4.0",
```
Comment (tnightengale) [inline build.sbt:277]: Hey @xx-db 👋 I saw you tried to add the mysql jdbc driver back in #211, but ultimately it wasn't merged. Currently the CI is complaining about a license issue by adding it here. Was that what you encountered? Do you know who might be able to either:

a) Approve overriding the license lint, and allow the connector to be used
b) Confirm that we can't use the connector due to a license conflict, and therefore the project should stop mentioning using MySQL

Thanks in advance!

CC: @dennyglee

File: Dockerfile [hunk 3]
```diff
@@ -1,147 +1,59 @@
...
+ENV HOME=$HOME \
+    JAVA_HOME=/usr/lib/jvm/default-jvm
 
-# Define the shell used within the container
-SHELL ["/bin/bash", "-i", "-c", "-o", "pipefail"]
+# Copy Java from base
+COPY --from=base $JAVA_HOME $JAVA_HOME
 
-ENV UC_SERVER_BIN="${unitycatalog_home}/${unitycatalog_bin}/start-uc-server"
+ENV PATH="${BASH_HOME}:${JAVA_HOME}/bin:${PATH}"
```
Comment (tnightengale) [inline Dockerfile:30]:
```suggestion
ENV PATH="${JAVA_HOME}/bin:${PATH}"
```

File: ui/Dockerfile (hunk 1 above)
Comment (jamieknight-db) [reply ui/Dockerfile:15]: @tnightengale to be honest, I am unsure the best way to parameterize this. Ideally, all parameterized variables are stored in the `.env` file and could be accessed from the package.json scripts. 

```
   "proxy": "${REACT_APP_API_URL}"
```

Or we might need to utilize a setupProxy.js file and createProxyMiddleware and go that route? 

Maybe @bijanhoule or @creechy or @yc-shawn can chime in?

File: README.md [hunk 4]
```diff
@@ -158,7 +163,40 @@ yarn start
...
+
+### Run with Postgres Container
+To run with a Postgres db container, instead of the default hibernate filesystem datastore, use the `compose.psql.yaml` file to [override](https://docs.docker.com/compose/multiple-compose-files/merge/#example) the default `compose.yaml` file:
+
+```sh
+docker compose -f compose.yaml -f compose.psql.yaml up -d
+```
 
+### Run with External Database
+The Unity Catalog container can be configured to connect with an external Postgres or MySQL database. Pass environment variables to the container to configure the [hibernate.properties](etc/conf/hibernate.properties) file. For example, `docker run --rm --env-file .env unitycatalog_backend_1`. See [.env.example](.env.example) for the possible variables.
```
Comment (tnightengale) [inline README.md:199]: TODO: get clarity on MySQL jdbc support inside the image.

File: server/src/main/java/io/unitycatalog/server/persist/utils/HibernateUtils.java [hunk 5]
```diff
@@ -31,10 +30,9 @@ private static SessionFactory createSessionFactory() {
         throw new RuntimeException("PropertiesUtil instance is null in createSessionFactory");
       }
 
-      Properties hibernateProperties = new Properties();
-      try (InputStream input = Files.newInputStream(Paths.get("etc/conf/hibernate.properties"))) {
-        hibernateProperties.load(input);
-      }
+      // Load properties with environment variable substitution to configure service in Docker
+      Properties hibernateProperties =
+          ConfigLoader.loadProperties(Paths.get("etc/conf/hibernate.properties"));
```
Comment (tnightengale) [inline server/src/main/java/io/unitycatalog/server/persist/utils/HibernateUtils.java:35]: Looks like this utility to read `.properties` files with env var substitution relies on https://github.com/spotbugs/spotbugs, which has a  LGPL-2.1 license. This is [failing the ci[(https://github.com/unitycatalog/unitycatalog/actions/runs/10932397607/job/30349161846).

Is this really prohibited?

File: Dockerfile [hunk 6]
```diff
@@ -1,147 +1,60 @@
...
-# Note: The default allocated heam memory size is too small
-# and will cause the process to fail when attempting to compile
-# and generate the Uber-Jar. Therefore it is important to choose
-# a size large enough for the compiler to run. 
-ARG sbt_args="-J-Xmx2G"
-# FIXME Pass it from the outside
-ARG unitycatalog_version="0.2.0-SNAPSHOT"
-ARG jars_directory="server/target/jars"
+# syntax=docker.io/docker/dockerfile:1.7-labs
+ARG HOME="/unitycatalog"
```
Comment (jaceklaskowski) [inline Dockerfile:2]: Why not stick to `/opt/unitycatalog`. It's shorter yet we don't really care, do we? `/opt` looks as the right place to me.

File: README.md [hunk 7]
```diff
@@ -158,7 +163,40 @@ yarn start
 open localhost:3000
 ```
 
+## Docker Quickstart
+To start Unity Catalog in Docker Compose in one command, install the latest version of [Docker Desktop](https://www.docker.com/products/docker-desktop/) and run the following:
```
Comment (jaceklaskowski) [inline README.md:167]: nit: Add a new line to separate the title from the text.

File: README.md [hunk 8]
```diff
@@ -158,7 +163,40 @@ yarn start
...
+
+```sh
+docker compose up -d
+```
+
+This starts the Unity Catalog UI and backend. You can access the UI at http://localhost:3000 and the backend at http://localhost:8080.
```
Comment (jaceklaskowski) [inline README.md:173]: This "backend" caught my attention. Is this "the Unity Catalog server" perhaps? What else?

You may want to write "all the Unity Catalog components" and link it to the respective architecture page where they're described (if there's any = no idea).

File: README.md [hunk 9]
```diff
@@ -158,7 +163,40 @@ yarn start
...
+
+To use the Unity Catalog CLI, attach to a shell in the backend container:
+
+```sh
+docker exec -it unitycatalog-backend-1 /bin/bash
```
Comment (jaceklaskowski) [inline README.md:178]: Oh, no, I'm sure that `unitycatalog-backend-1` will soon be the number 1 for many to replace with something more meaningful. Can you replace `backend-1` and likely make `unitycatalog` shorter (say `uc`)? 🙏

File: build.sbt [hunk 10]
```diff
@@ -270,6 +271,15 @@ lazy val server = (project in file("server"))
       "com.h2database" %  "h2" % "2.2.224",
 
       "org.hibernate.orm" % "hibernate-core" % "6.5.0.Final",
+      // For loading properties from the environment
+      "uk.org.webcompere" % "lightweight-config" % "1.3.0",
+
+      // TODO: Determine if project will support MySQL jar; linter currently states
```
Comment (jaceklaskowski) [inline build.sbt:277]: Remove it until it's requested and worked out.

File: build.sbt [hunk 11]
```diff
@@ -270,6 +271,15 @@ lazy val server = (project in file("server"))
...
+      // license is imcompatible
+      // "com.mysql" % "mysql-connector-j" % "8.4.0",
+
+      // For PostgreSQL backend
+      "org.postgresql" % "postgresql" % "42.6.0",
```
Comment (jaceklaskowski) [inline build.sbt:282]: I don't think we want this to be included by default. It's one of the many hibernate-supported databases and no need to make this any better than others (but including it in the build). UC does not need it, correct? If so, whoever wants PostgreSQL, they'll download the jars and make necessary changes. WDYT?

File: compose.yaml [hunk 12]
```diff
@@ -0,0 +1,25 @@
+name: unitycatalog
+
+services:
+
+  backend:
```
Comment (jaceklaskowski) [inline compose.yaml:5]: s/backend/server

File: README.md [hunk 13]
```diff
@@ -65,8 +72,6 @@ You have to ensure that your local environment has the following:
 - Ensure the `JAVA_HOME` environment variable your terminal is configured to point to JDK17.
 - Compile the project using `build/sbt package`
```
Comment (jaceklaskowski) [inline README.md:73]: Is this still required?

**Comment by jaceklaskowski** [comment]
`docker/README.md` needs some love too, doesn't it? Can you review it and make it up-to-date?

**Comment by jaceklaskowski** [comment]
I couldn't build the images atm as it breaks on pulling delta-spark 3.2.1 😢  I'll do it again once all the comments are resolved.

```
#12 170.5 [error] coursier.ResolutionException: Encountered 1 error(s) in dependency resolution:
#12 170.5 [error]     io.delta:delta-spark_2.12:3.2.1:
#12 170.5 [error]         not found:
```

File: README.md (hunk 8 above)
Comment (tnightengale) [reply README.md:173]: To clarify, does `./bin/start-uc-server` start all the components? Because that is the CMD for the image.

File: README.md (hunk 9 above)
Comment (tnightengale) [reply README.md:178]: This is default from Docker Compose, which creates containers according to the following: <docker-compose-name>_<service-name>_<service_container_cardinatlity>.

I think it's better to give the complete name: `unitycatalog` and be more clear about the container throughout its lifecycle, than to use shorthands.

File: build.sbt (hunk 11 above)
Comment (tnightengale) [reply build.sbt:282]: I think there's arguments for both.

Considering that a real database is required to run UC in production, I think it might be good practice to include the supported backends officially, by adding them as build dependencies. It would massively simplify using their respective JDBC jars, rather than trying to mount them to the container AND update the classpath.

We could also then _test_ the backends, using docker-compose with each database. I think this relates to my earlier comments about what databases are considering "officially supported", if any.

The alternative of not including the jars still requires the project to implicitly support different databases, but with less control over testing them. Which in turn, could lead to more questions from the community, because to "official" pattern to self-hosting UC in production is less defined.

For example, not including the jars would require us to modify the image's entrypoint to handle arbitrary ENV vars for classpath modifications and require the user to correctly mount the JDBC jar. 

In my opinion, there's no need to support and document that additional interface. It would be quite brittle, and basically untested. Which means it is not as reliable for self-hosting in production, compared to using a "officially supported" and included jar.

To summarize, both approaches have merit, but I think we should hear from other maintainers if the database interface should have "officially supported" connectors.

File: compose.yaml (hunk 12 above)
Comment (tnightengale) [reply compose.yaml:5]: Can you clarify this?

**Comment by tnightengale** [comment]
> `docker/README.md` needs some love too, doesn't it? Can you review it and make it up-to-date?

Basically I think we could deprecate and remove `./docker` with this PR. Thoughts?

File: build.sbt (hunk 2 above)
Comment (xx-db) [reply build.sbt]: @tnightengale 👋You are right, it was ultimately not merged because they are commonly used jars supported by Hibernate. So we included documentation to download the connector jars instead.

File: build.sbt (hunk 11 above)
Comment (tnightengale) [reply build.sbt:282]: @tdas Can we get some guidance from yourself or another maintainer on if UC should offer support for "official" backends/databases and include those JDBC connector jars in the forthcoming official image? 

I think I see two paths here. Open to any thoughts on other directions as well.

**1. UC does not include "officially supported" databases to persist its application data. Users of the official image would be responsible for the following:**
a. Mount the correct JDBC jars to the container
b. Modify the relevant `classpath` to reference those jars
c. Mount their own `./etc` config files for hibernate to the container

**Pros**: Smaller image, less project dependencies.
**Cons**: Allows ambiguity about the "best" approach to self-hosting with an external database in production. No testing of different connections and a brittle, implicit interface, which means we could break things that users rely on. For example, reorganizing the config file directories would break a users mount in their deployment scripts.

**2. UC includes "officially supported" JDBC jars (eg. MySQL & Postgres) in the official image. Users of the image would then:**
a. configure their connection by passing env vars to the container (possible vars can be added to `.properties` config files with defaults and interpolated in the app OR
b. Mount their own `./etc` config files for hibernate to the container

**Pros**: We could test the different officially supported databases in CI using docker compose. 
**Cons**: Additional jars means a slightly bigger image size for folks who would only use one of the supported databases.

File: README.md (hunk 8 above)
Comment (jaceklaskowski) [reply README.md:173]: All the default components of the UC Localhost Server, yes. I think this backend term comes more from Docker Compose where it's used. Correct?

File: README.md (hunk 9 above)
Comment (jaceklaskowski) [reply README.md:178]: Let's rename `backend` to be `server` to give more clarity.

File: compose.yaml (hunk 12 above)
Comment (jaceklaskowski) [reply compose.yaml:5]: I'd much appreciate if you could rename `backend` with `server` to give this service a bit more UC-specific flavour.

File: README.md (hunk 13 above)
Comment (jaceklaskowski) [reply README.md:73]: Is `build/sbt package` required to have a dockerized environment? I don't think so as `sbt package` is part of building the UC images.

Comment (dennyglee) [reply README.md:73]: Having the `build/sbt` package would be helpful for anyone building apps though.  Not a blocker though.

**Comment by dennyglee** [comment]
> I couldn't build the images atm as it breaks on pulling delta-spark 3.2.1 😢 I'll do it again once all the comments are resolved.
> 
> ```
> #12 170.5 [error] coursier.ResolutionException: Encountered 1 error(s) in dependency resolution:
> #12 170.5 [error]     io.delta:delta-spark_2.12:3.2.1:
> #12 170.5 [error]         not found:
> ```

Yeah, that should be soon :)

**Comment by dennyglee** [comment]
> > `docker/README.md` needs some love too, doesn't it? Can you review it and make it up-to-date?
> 
> Is there any reason to keep `./docker/` given the simplifications in this PR and the updated `./README.md`?

Possibly though I'm not personally hard set on this.  One of the reasons it *may* make sense is in the past it was difficult to get the docker to run in DockerHub unless it had its own folder and/or root repository.  This may have changed and/or perhaps I configured it wrong so I would glad to be corrected here, eh?!

**Comment by jaceklaskowski** [comment]
I'd keep README.md's separate as the main one gets fairly large and should really be part of the official docs (under `docs` ofc). I'd simply leave a note about Unity Catalog in Docker in the main `README.md` with a link to the `docker/README.md`.

I'm re-wording the main `README.md` as we speak (so perhaps it's going to be as tiny that the other `docker/README.md` becomes useless). We'll see.

**Comment by jaceklaskowski** [comment]
Pulled the PR and reviewed the changes, and I'm very very sure that we should keep all docker-related files in their own `docker` directory. Having a dockerized environment is very useful, but that's not the main goal of Unity Catalog and is simply a convenience.

That's why I'm requesting the following files to be in `docker` directory:

1. compose.psql.yaml
2. compose.yaml
3. Dockerfile

(perhaps even `ui/Dockerfile` to keep everything docker-related in one directory).

**Comment by jaceklaskowski** [comment]
@dennyglee @tnightengale Got a question about how to contribute to the PR. I must admit that I've never done it before (to contribute to a PR with a PR) and...could you guide me how to do it properly?

I asked the same question on slack --> https://unitycatalog.slack.com/archives/C076YREKX8W/p1727615205499499?thread_ts=1727102843.375399&cid=C076YREKX8W

**Comment by dennyglee** [comment]
> @dennyglee @tnightengale Got a question about how to contribute to the PR. I must admit that I've never done it before (to contribute to a PR with a PR) and...could you guide me how to do it properly?
> 
> I asked the same question on slack --> https://unitycatalog.slack.com/archives/C076YREKX8W/p1727615205499499?thread_ts=1727102843.375399&cid=C076YREKX8W

Oh generally, I create a fork of the OPR (original PR) and then submit the PR to that one.  e.g., 
- main is `unitycatalog/unitycatalog` branch `main`
- Original PR is `foo/unitycatalog` branch `bar`
- New PR is `blah/unitycatalog` branch `bar` where this PR is submitted to `foo/unitycatalog`

**Comment by tnightengale** [comment]
> > > `docker/README.md` needs some love too, doesn't it? Can you review it and make it up-to-date?
> > 
> > 
> > Is there any reason to keep `./docker/` given the simplifications in this PR and the updated `./README.md`?
> 
> Possibly though I'm not personally hard set on this. One of the reasons it _may_ make sense is in the past it was difficult to get the docker to run in DockerHub unless it had its own folder and/or root repository. This may have changed and/or perhaps I configured it wrong so I would glad to be corrected here, eh?!

@dennyglee To clarify, what do you mean by 'run in Dockerhub'? As I understand it, DockerHub is just an image registry, like any other, and we would just be pushing a built image artifact to it.

On the topic of keeping everything in `./docker/*`, I would suggest we keep the files in the `unitycatalog/` project: this is the standard for most official images because it keeps things simple. The image is an artifact of the whole project; not a subcomponent.

**Comment by tnightengale** [comment]
@dennyglee @xx-db I decoupled any MySQL/Postgres dependencies from the image.

This PR now responsible has a few tightly coupled responsibilities:
1. Unifying the multiple Dockerfiles into one image
2. Deprecating the previous Docker docs
3. Creating a simple interface to run UC in Docker, using `docker compose up`
4. Adding a standard Github Actions workflow to build the image

File: README.md [hunk 14]
```diff
@@ -57,6 +57,9 @@ Let's take Unity Catalog for spin. In this guide, we are going to do the followi
   An example project is provided to demonstrate how to use the UC SDK for various assets
   as well as provide a convenient way to explore the content of any UC server implementation.
 
+> If you prefer to run Unity Catalog in Docker use `docker
+> compose up`. See the [Docker Quickstart](./docs/docker_compose_quickstart.md) for more details.
```
Comment (tnightengale) [inline README.md:61]:
```suggestion
> compose up`. See the [Docker Quickstart](./docs/docker_quickstart.md) for more details.
```

**Comment by tnightengale** [comment]
@dennyglee Hey, rewriting commit history (see above) using the suggested [DCO](https://github.com/unitycatalog/unitycatalog/pull/485/checks?check_run_id=30920376331) commands, `git rebase HEAD~43 --signoff` and `git push --force-with-lease origin teghan/build_one_dockerfile` didn't seem to work and has left the history of this branch a hot mess of garbage.

How do you feel about me opening a new branch with one new signed and squashed commit containing the changes here?
